### ISSUE TITLE
🐛 Bug Report: textextensions module not found error when I use @backstage/backend-test-utils

### ISSUE THREAD

**Issue by RaoJackie123** [backstage/backstage#26341]
### 📜 Description

When I run test, and get below error:

  ● Test suite failed to run

    Cannot find module 'textextensions' from '../../../../node_modules/@backstage/backend-test-utils/dist/index.cjs.js'

    Require stack:
      /home/runner/work/roadie-backstage-plugins/roadie-backstage-plugins/node_modules/@backstage/backend-test-utils/dist/index.cjs.js
      actions/run/backstageRequest.test.ts

### 👍 Expected behavior

The test should work fine

### 👎 Actual Behavior with Screenshots

![image](https://github.com/user-attachments/assets/a4ef29d5-511e-45b4-ad5f-0ef86d2d0258)
![image](https://github.com/user-attachments/assets/6e9e2252-5cbd-4ab6-a8fb-ed3360b27bcd)

### 👟 Reproduction steps

Here's the CI job result:

https://github.com/RoadieHQ/roadie-backstage-plugins/actions/runs/10588590345/job/29350483549

here's my code:
https://github.com/RoadieHQ/roadie-backstage-plugins/pull/1532/files#diff-36c0042fb2aff216f63d66aed0c18a5be142206daddcff4be556d1e17142db21R22

I just use the mockServices, mockCredentials in my test script.

### 🖥️ Your Environment

OS:   Windows_NT 10.0.19045 - win32/x64
node: v18.20.4
yarn: 1.22.19
cli:  0.27.0 (installed)
backstage:  1.30.2

Dependencies:
  @backstage/app-defaults                                          1.5.10
  @backstage/backend-app-api                                       0.9.2
  @backstage/backend-common                                        0.24.0
  @backstage/backend-defaults                                      0.4.3
  @backstage/backend-dev-utils                                     0.1.5
  @backstage/backend-openapi-utils                                 0.1.16
  @backstage/backend-plugin-api                                    0.8.0
  @backstage/backend-tasks                                         0.6.0
  @backstage/backend-test-utils                                    0.5.0
  @backstage/catalog-client                                        1.6.6
  @backstage/catalog-model                                         1.6.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.7
  @backstage/cli                                                   0.27.0
  @backstage/config-loader                                         1.9.0
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.14.2
  @backstage/core-compat-api                                       0.2.8
  @backstage/core-components                                       0.14.10
  @backstage/core-plugin-api                                       1.9.3
  @backstage/dev-utils                                             1.0.37
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.8
  @backstage/frontend-app-api                                      0.8.0
  @backstage/frontend-plugin-api                                   0.6.4, 0.7.0
  @backstage/frontend-test-utils                                   0.1.12
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.30
  @backstage/integration                                           1.14.0
  @backstage/plugin-api-docs                                       0.11.8
  @backstage/plugin-app-backend                                    0.3.72
  @backstage/plugin-app-node                                       0.1.23
  @backstage/plugin-auth-backend-module-atlassian-provider         0.2.4
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.15
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.1.6
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.1.6
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.2.0
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.18
  @backstage/plugin-auth-backend-module-github-provider            0.1.20
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.20
  @backstage/plugin-auth-backend-module-google-provider            0.1.20
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.18
  @backstage/plugin-auth-backend-module-oauth2-provider            0.2.4
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.16
  @backstage/plugin-auth-backend-module-oidc-provider              0.2.4
  @backstage/plugin-auth-backend-module-okta-provider              0.0.16
  @backstage/plugin-auth-backend-module-onelogin-provider          0.1.4
  @backstage/plugin-auth-backend                                   0.22.10
  @backstage/plugin-auth-node                                      0.5.0
  @backstage/plugin-auth-react                                     0.1.5
  @backstage/plugin-bitbucket-cloud-common                         0.2.22
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.21
  @backstage/plugin-catalog-backend                                1.25.0
  @backstage/plugin-catalog-common                                 1.0.26
  @backstage/plugin-catalog-graph                                  0.4.8
  @backstage/plugin-catalog-import                                 0.12.2
  @backstage/plugin-catalog-node                                   1.12.5
  @backstage/plugin-catalog-react                                  1.12.3
  @backstage/plugin-catalog                                        1.22.0
  @backstage/plugin-events-node                                    0.3.9
  @backstage/plugin-home-react                                     0.1.16
  @backstage/plugin-home                                           0.7.9
  @backstage/plugin-org                                            0.6.28
  @backstage/plugin-permission-common                              0.8.1
  @backstage/plugin-permission-node                                0.8.1
  @backstage/plugin-permission-react                               0.4.25
  @backstage/plugin-proxy-backend                                  0.5.4
  @backstage/plugin-scaffolder-backend-module-azure                0.1.15
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.13
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.13
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.13
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.15
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.13
  @backstage/plugin-scaffolder-backend-module-github               0.4.1
  @backstage/plugin-scaffolder-backend-module-gitlab               0.4.5
  @backstage/plugin-scaffolder-backend                             1.24.0
  @backstage/plugin-scaffolder-common                              1.5.5
  @backstage/plugin-scaffolder-node                                0.4.9
  @backstage/plugin-scaffolder-react                               1.11.0
  @backstage/plugin-scaffolder                                     1.24.0
  @backstage/plugin-search-backend-module-catalog                  0.2.0
  @backstage/plugin-search-backend-module-techdocs                 0.2.0
  @backstage/plugin-search-backend-node                            1.3.0
  @backstage/plugin-search-common                                  1.2.14
  @backstage/plugin-search-react                                   1.7.14
  @backstage/plugin-search                                         1.4.15
  @backstage/plugin-signals-react                                  0.0.4
  @backstage/plugin-techdocs-backend                               1.10.11
  @backstage/plugin-techdocs-common                                0.1.0
  @backstage/plugin-techdocs-node                                  1.12.9
  @backstage/plugin-techdocs-react                                 1.2.7
  @backstage/plugin-techdocs                                       1.10.8
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.11
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.5.10
  @backstage/theme                                                 0.5.6
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.8

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

**Comment by RaoJackie123** [comment]
It's similar to https://github.com/backstage/backstage/issues/26197, but this should be a different thing.

**Comment by benjdlambert** [comment]
I'm not sure where this `testextensions` is required, I don't see it anywhere in our code. Does the `dist/index.cjs` contain `testextensions`?

**Comment by benjdlambert** [comment]
@RaoJackie123 Sorry, I see the issue now. I misread `testextensions` and it's `textextensions`. I think that this is the same issue as the `pg-format` one. I

**Comment by RaoJackie123** [comment]
> @RaoJackie123 ~Sorry, I see the issue now. I misread `testextensions` and it's `textextensions`. I think that this is the same issue as the `pg-format` one~
> 
> Hmm this is similar, but I don't think it's the same.
> 
> Could you try with `1.30.4` and make sure that dependencies are de-duped? `yarn why @backstage/backend-test-utils` should help to see if you have duplicates.

Do you mean upgrade the backstage to 1.30.4 will solve it?

**Comment by RaoJackie123** [comment]
> yarn backstage-cli versions:bump

I just upgrade the backstage in my local, but still get the same issue, the env like below:

OS:   Windows_NT 10.0.19045 - win32/x64
node: v18.20.4
yarn: 1.22.19
cli:  0.27.0 (installed)
backstage:  1.30.4

Dependencies:
  @backstage/app-defaults                                          1.5.10
  @backstage/backend-app-api                                       0.9.3
  @backstage/backend-common                                        0.24.1
  @backstage/backend-defaults                                      0.4.4
  @backstage/backend-dev-utils                                     0.1.5
  @backstage/backend-openapi-utils                                 0.1.17
  @backstage/backend-plugin-api                                    0.8.1
  @backstage/backend-tasks                                         0.6.1
  @backstage/backend-test-utils                                    0.5.1
  @backstage/catalog-client                                        1.6.6
  @backstage/catalog-model                                         1.6.0
  @backstage/cli-common                                            0.1.14
  @backstage/cli-node                                              0.2.7
  @backstage/cli                                                   0.27.0
  @backstage/config-loader                                         1.9.0
  @backstage/config                                                1.2.0
  @backstage/core-app-api                                          1.14.2
  @backstage/core-compat-api                                       0.2.8
  @backstage/core-components                                       0.14.10
  @backstage/core-plugin-api                                       1.9.3
  @backstage/dev-utils                                             1.0.37
  @backstage/errors                                                1.2.4
  @backstage/eslint-plugin                                         0.1.8
  @backstage/frontend-app-api                                      0.8.0
  @backstage/frontend-plugin-api                                   0.6.4, 0.7.0
  @backstage/frontend-test-utils                                   0.1.12
  @backstage/integration-aws-node                                  0.1.12
  @backstage/integration-react                                     1.1.30
  @backstage/integration                                           1.14.0
  @backstage/plugin-api-docs                                       0.11.8
  @backstage/plugin-app-backend                                    0.3.73
  @backstage/plugin-app-node                                       0.1.24
  @backstage/plugin-auth-backend-module-atlassian-provider         0.2.5
  @backstage/plugin-auth-backend-module-aws-alb-provider           0.1.17
  @backstage/plugin-auth-backend-module-azure-easyauth-provider    0.1.7
  @backstage/plugin-auth-backend-module-bitbucket-provider         0.1.7
  @backstage/plugin-auth-backend-module-cloudflare-access-provider 0.2.1
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.19
  @backstage/plugin-auth-backend-module-github-provider            0.1.21
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.21
  @backstage/plugin-auth-backend-module-google-provider            0.1.21
  @backstage/plugin-auth-backend-module-microsoft-provider         0.1.19
  @backstage/plugin-auth-backend-module-oauth2-provider            0.2.5
  @backstage/plugin-auth-backend-module-oauth2-proxy-provider      0.1.17
  @backstage/plugin-auth-backend-module-oidc-provider              0.2.6
  @backstage/plugin-auth-backend-module-okta-provider              0.0.17
  @backstage/plugin-auth-backend-module-onelogin-provider          0.1.5
  @backstage/plugin-auth-backend                                   0.22.12
  @backstage/plugin-auth-node                                      0.5.1
  @backstage/plugin-auth-react                                     0.1.5
  @backstage/plugin-bitbucket-cloud-common                         0.2.22
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.22
  @backstage/plugin-catalog-backend                                1.25.2
  @backstage/plugin-catalog-common                                 1.0.26
  @backstage/plugin-catalog-graph                                  0.4.8
  @backstage/plugin-catalog-import                                 0.12.2
  @backstage/plugin-catalog-node                                   1.12.6
  @backstage/plugin-catalog-react                                  1.12.3
  @backstage/plugin-catalog                                        1.22.0
  @backstage/plugin-events-node                                    0.3.10
  @backstage/plugin-home-react                                     0.1.16
  @backstage/plugin-home                                           0.7.9
  @backstage/plugin-org                                            0.6.28
  @backstage/plugin-permission-common                              0.8.1
  @backstage/plugin-permission-node                                0.8.2
  @backstage/plugin-permission-react                               0.4.25
  @backstage/plugin-proxy-backend                                  0.5.5
  @backstage/plugin-scaffolder-backend-module-azure                0.1.16
  @backstage/plugin-scaffolder-backend-module-bitbucket-cloud      0.1.14
  @backstage/plugin-scaffolder-backend-module-bitbucket-server     0.1.14
  @backstage/plugin-scaffolder-backend-module-bitbucket            0.2.14
  @backstage/plugin-scaffolder-backend-module-gerrit               0.1.16
  @backstage/plugin-scaffolder-backend-module-gitea                0.1.14
  @backstage/plugin-scaffolder-backend-module-github               0.4.2
  @backstage/plugin-scaffolder-backend-module-gitlab               0.4.6
  @backstage/plugin-scaffolder-backend                             1.24.1
  @backstage/plugin-scaffolder-common                              1.5.5
  @backstage/plugin-scaffolder-node                                0.4.10
  @backstage/plugin-scaffolder-react                               1.11.0
  @backstage/plugin-scaffolder                                     1.24.0
  @backstage/plugin-search-backend-module-catalog                  0.2.1
  @backstage/plugin-search-backend-module-techdocs                 0.2.1
  @backstage/plugin-search-backend-node                            1.3.1
  @backstage/plugin-search-common                                  1.2.14
  @backstage/plugin-search-react                                   1.7.14
  @backstage/plugin-search                                         1.4.15
  @backstage/plugin-signals-react                                  0.0.4
  @backstage/plugin-techdocs-backend                               1.10.12
  @backstage/plugin-techdocs-common                                0.1.0
  @backstage/plugin-techdocs-node                                  1.12.10
  @backstage/plugin-techdocs-react                                 1.2.7
  @backstage/plugin-techdocs                                       1.10.8
  @backstage/plugin-user-settings-common                           0.0.1
  @backstage/plugin-user-settings                                  0.8.11
  @backstage/release-manifests                                     0.0.11
  @backstage/test-utils                                            1.5.10
  @backstage/theme                                                 0.5.6
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.8

**Comment by RaoJackie123** [comment]
> I'm not confident, but it could be worth a try. It could be something to do with some dependencies being hoisted in the wrong place. So `yarn backstage-cli versions:bump && yarn dedupe` or `npx yarn-deduplicate` if you're running older versions of `yarn`

I just upgrade the backstage in my local, but still get the same issue, the env like below:

OS: Windows_NT 10.0.19045 - win32/x64
node: v18.20.4
yarn: 1.22.19
cli: 0.27.0 (installed)
backstage: 1.30.4

Dependencies:
@backstage/app-defaults 1.5.10
@backstage/backend-app-api 0.9.3
@backstage/backend-common 0.24.1
@backstage/backend-defaults 0.4.4
@backstage/backend-dev-utils 0.1.5
@backstage/backend-openapi-utils 0.1.17
@backstage/backend-plugin-api 0.8.1
@backstage/backend-tasks 0.6.1
@backstage/backend-test-utils 0.5.1
@backstage/catalog-client 1.6.6
@backstage/catalog-model 1.6.0
@backstage/cli-common 0.1.14
@backstage/cli-node 0.2.7
@backstage/cli 0.27.0
@backstage/config-loader 1.9.0
@backstage/config 1.2.0
@backstage/core-app-api 1.14.2
@backstage/core-compat-api 0.2.8
@backstage/core-components 0.14.10
@backstage/core-plugin-api 1.9.3
@backstage/dev-utils 1.0.37
@backstage/errors 1.2.4
@backstage/eslint-plugin 0.1.8
@backstage/frontend-app-api 0.8.0
@backstage/frontend-plugin-api 0.6.4, 0.7.0
@backstage/frontend-test-utils 0.1.12
@backstage/integration-aws-node 0.1.12
@backstage/integration-react 1.1.30
@backstage/integration 1.14.0
@backstage/plugin-api-docs 0.11.8
@backstage/plugin-app-backend 0.3.73
@backstage/plugin-app-node 0.1.24
@backstage/plugin-auth-backend-module-atlassian-provider 0.2.5
@backstage/plugin-auth-backend-module-aws-alb-provider 0.1.17
@backstage/plugin-auth-backend-module-azure-easyauth-provider 0.1.7
@backstage/plugin-auth-backend-module-bitbucket-provider 0.1.7
@backstage/plugin-auth-backend-module-cloudflare-access-provider 0.2.1
@backstage/plugin-auth-backend-module-gcp-iap-provider 0.2.19
@backstage/plugin-auth-backend-module-github-provider 0.1.21
@backstage/plugin-auth-backend-module-gitlab-provider 0.1.21
@backstage/plugin-auth-backend-module-google-provider 0.1.21
@backstage/plugin-auth-backend-module-microsoft-provider 0.1.19
@backstage/plugin-auth-backend-module-oauth2-provider 0.2.5
@backstage/plugin-auth-backend-module-oauth2-proxy-provider 0.1.17
@backstage/plugin-auth-backend-module-oidc-provider 0.2.6
@backstage/plugin-auth-backend-module-okta-provider 0.0.17
@backstage/plugin-auth-backend-module-onelogin-provider 0.1.5
@backstage/plugin-auth-backend 0.22.12
@backstage/plugin-auth-node 0.5.1
@backstage/plugin-auth-react 0.1.5
@backstage/plugin-bitbucket-cloud-common 0.2.22
@backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.22
@backstage/plugin-catalog-backend 1.25.2
@backstage/plugin-catalog-common 1.0.26
@backstage/plugin-catalog-graph 0.4.8
@backstage/plugin-catalog-import 0.12.2
@backstage/plugin-catalog-node 1.12.6
@backstage/plugin-catalog-react 1.12.3
@backstage/plugin-catalog 1.22.0
@backstage/plugin-events-node 0.3.10
@backstage/plugin-home-react 0.1.16
@backstage/plugin-home 0.7.9
@backstage/plugin-org 0.6.28
@backstage/plugin-permission-common 0.8.1
@backstage/plugin-permission-node 0.8.2
@backstage/plugin-permission-react 0.4.25
@backstage/plugin-proxy-backend 0.5.5
@backstage/plugin-scaffolder-backend-module-azure 0.1.16
@backstage/plugin-scaffolder-backend-module-bitbucket-cloud 0.1.14
@backstage/plugin-scaffolder-backend-module-bitbucket-server 0.1.14
@backstage/plugin-scaffolder-backend-module-bitbucket 0.2.14
@backstage/plugin-scaffolder-backend-module-gerrit 0.1.16
@backstage/plugin-scaffolder-backend-module-gitea 0.1.14
@backstage/plugin-scaffolder-backend-module-github 0.4.2
@backstage/plugin-scaffolder-backend-module-gitlab 0.4.6
@backstage/plugin-scaffolder-backend 1.24.1
@backstage/plugin-scaffolder-common 1.5.5
@backstage/plugin-scaffolder-node 0.4.10
@backstage/plugin-scaffolder-react 1.11.0
@backstage/plugin-scaffolder 1.24.0
@backstage/plugin-search-backend-module-catalog 0.2.1
@backstage/plugin-search-backend-module-techdocs 0.2.1
@backstage/plugin-search-backend-node 1.3.1
@backstage/plugin-search-common 1.2.14
@backstage/plugin-search-react 1.7.14
@backstage/plugin-search 1.4.15
@backstage/plugin-signals-react 0.0.4
@backstage/plugin-techdocs-backend 1.10.12
@backstage/plugin-techdocs-common 0.1.0
@backstage/plugin-techdocs-node 1.12.10
@backstage/plugin-techdocs-react 1.2.7
@backstage/plugin-techdocs 1.10.8
@backstage/plugin-user-settings-common 0.0.1
@backstage/plugin-user-settings 0.8.11
@backstage/release-manifests 0.0.11
@backstage/test-utils 1.5.10
@backstage/theme 0.5.6
@backstage/types 1.1.1
@backstage/version-bridge 1.0.8

**Comment by ScottGuymer** [comment]
I also seeing the same issue when adopting `mockServices` from '@backstage/backend-test-utils'.

I'm trying to remove the deprecated `Logger` dependencies.

**Comment by ScottGuymer** [comment]
I also seeing the same issue when adopting `mockServices` from '@backstage/backend-test-utils'.

I'm trying to remove the deprecated `Logger` dependencies.

I can see that `textextensions` is used here https://github.com/backstage/backstage/blob/8f85ab3b04e3311171ca281d722c5177a5bba576/packages/backend-test-utils/src/filesystem/MockDirectory.ts#L20

**Comment by benjdlambert** [comment]
@ScottGuymer but it's listed in the dependencies so I'm not sure how it's missing. Can you get me a full stack trace of the error?

**Comment by ScottGuymer** [comment]
Yeah it's a very odd one. I checked through what actually gets installed and it seems to be there.

This is what I get out of jest

```
  ● Test suite failed to run

    Cannot find module 'textextensions' from '../../../node_modules/@backstage/backend-test-utils/dist/index.cjs.js'

    Require stack:
      /workspaces/developer-portal/backstage/node_modules/@backstage/backend-test-utils/dist/index.cjs.js
      findProjectByKey.test.ts

       5 | import { MockConfigApi, setupRequestMockHandlers } from '@backstage/test-utils';
       6 | import { setupServer } from 'msw/node';
    >  7 |
         | ^
       8 | const singleProject = {
       9 |   paging: {
      10 |     pageIndex: 1,

      at Resolver._throwModNotFoundError (../../../node_modules/jest-resolve/build/resolver.js:427:11)
      at Object.require (../../../node_modules/@backstage/backend-test-utils/dist/index.cjs.js:15:22)
      at Object.<anonymous> (findProjectByKey.test.ts:7:27)
```

**Comment by benjdlambert** [comment]
@ScottGuymer  Hmm, I just checked their repo and I came across this: https://github.com/bevry/textextensions/pull/251/files

I wonder if you can make the `package.json` change locally from that PR and test it?

Otherwise what versions of `jest-resolve` are you using, node version etc?

**Comment by ScottGuymer** [comment]
So i added a resolution in my base package.json to bump to `"textextensions": "^6.11.0"`` and then `yarn dedupe` (still broken at this point)

Made the changes in the PR to the local node_modules version and it had no effect.

```
$ yarn why jest-resolve
├─ @jest/core@npm:29.7.0
│  └─ jest-resolve@npm:29.7.0 (via npm:^29.7.0)
│
├─ @jest/core@npm:29.7.0 [a8395]
│  └─ jest-resolve@npm:29.7.0 (via npm:^29.7.0)
│
├─ jest-config@npm:29.7.0
│  └─ jest-resolve@npm:29.7.0 (via npm:^29.7.0)
│
├─ jest-config@npm:29.7.0 [3a6a7]
│  └─ jest-resolve@npm:29.7.0 (via npm:^29.7.0)
│
├─ jest-config@npm:29.7.0 [72361]
│  └─ jest-resolve@npm:29.7.0 (via npm:^29.7.0)
│
├─ jest-runner@npm:29.7.0
│  └─ jest-resolve@npm:29.7.0 (via npm:^29.7.0)
│
└─ jest-runtime@npm:29.7.0
   └─ jest-resolve@npm:29.7.0 (via npm:^29.7.0)
```

```
$ node --version
v18.20.0
```

**Comment by benjdlambert** [comment]
@ScottGuymer @RaoJackie123 is there any way either of you can put together a reproduction repo with a new `create-app` or something? These issues are always super tricky to debug without having access to a repo 😅

**Comment by benjdlambert** [comment]
OK, so I did some digging, and it look like it could perhaps be a `yarn@v1` bug. I could reproduce this locally with the branch provided from the roadie repo, and then bumped to `yarn@v4` and did a dedupe of dependencies and it seems to work. @ScottGuymer are you also using `yarn@1`?

**Comment by ScottGuymer** [comment]
Nope we are on yarn 4.

😬😬

**Comment by ScottGuymer** [comment]
Here is a repro repo

https://github.com/ScottGuymer/backstage-logger-repro

this one is using OG yarn

``
cd backstage
yarn install
yarn test
```

you should get 

```
 FAIL   @internal/backstage-plugin-example-common  plugins/example-common/src/example.test.ts
  ● Test suite failed to run

    Cannot find module 'textextensions' from '../../../node_modules/@backstage/backend-test-utils/dist/index.cjs.js'

    Require stack:
      /Users/scott/Code/backstage-logger-repro/backstage/node_modules/@backstage/backend-test-utils/dist/index.cjs.js
      example.test.ts

      3 | const logger = mockServices.logger.mock();
      4 |
    > 5 | describe('example', () => {
        |                           ^
      6 |   it('should pass', async () => {
      7 |     expect(logger).toBeDefined();
      8 |     expect(1).toBe(1);

      at Resolver._throwModNotFoundError (../../../node_modules/jest-resolve/build/resolver.js:427:11)
      at Object.<anonymous> (../../../node_modules/@backstage/backend-test-utils/dist/index.cjs.js:55:22)
      at Object.<anonymous> (example.test.ts:5:27)
```


While reproing this I see that it only happens in a plugin. And copying what i have locally this plugin is `"role": "common-library"`, not sure if this is having some kind of impact.

Identical tests in `backend` and the plugin, one pases and the other fails...

**Comment by benjdlambert** [comment]
AH @ScottGuymer you've hit the nail on the head. If you change the role in the `package.json` for the `plugin-example-common` package it starts to work. It's failing because you can't depend on `backend-test-utils` from the `common-library` package. We have [some tooling](https://github.com/backstage/backstage/blob/master/scripts/verify-local-dependencies.js) in the main repo which detects incompatible dependencies, that I think we should probably think of making a lint rule out of or something to help with these issues.

It does fail in a rather odd way, but I think it's because of the resolution `testextensions` failing under `.cjs` which happens when using `common-library`.

In short, because `backend-test-utils` is not polymorphic, you're gonna have to remove that dependency.

What are you both trying to use from `backend-test-utils` that you need in `common-library`? It might be that we should move some things out if there's reason to do so.

**Comment by ScottGuymer** [comment]
We are just using it to mock the logger service.

Tbh we can probably move the thing in this library to a node plugin. I'll have to look where it is used not sure that it's actually polymorphic

**Comment by ScottGuymer** [comment]
Taking this knowledge back to my backstage instance doesnt seem to help.

Changing the type in the package.json doesnt seem to have any affect and it keeps failing.

**Comment by Rugvip** [comment]
@RaoJackie123 are you still facing this issue, or is it alright to close?

**Comment by RaoJackie123** [comment]
> AH @ScottGuymer you've hit the nail on the head. If you change the role in the `package.json` for the `plugin-example-common` package it starts to work. It's failing because you can't depend on `backend-test-utils` from the `common-library` package. We have [some tooling](https://github.com/backstage/backstage/blob/master/scripts/verify-local-dependencies.js) in the main repo which detects incompatible dependencies, that I think we should probably think of making a lint rule out of or something to help with these issues.
> 
> It does fail in a rather odd way, but I think it's because of the resolution `testextensions` failing under `.cjs` which happens when using `common-library`.
> 
> In short, because `backend-test-utils` is not polymorphic, you're gonna have to remove that dependency.
> 
> What are you both trying to use from `backend-test-utils` that you need in `common-library`? It might be that we should move some things out if there's reason to do so.

I'm using it from 'backend-plugin-module', and just use mockServices to mock the auth and log
  "backstage": {
    "role": "backend-plugin-module"
  }

**Comment by RaoJackie123** [comment]
> @RaoJackie123 are you still facing this issue, or is it alright to close?

Hi @Rugvip Still can't find an effect way to solve the issue, it's working in yarn 4, but not working in yarn 1. You can close it temporally.

**Comment by grantila** [comment]
This is still an issue for `frontend-plugin` roles not being able to import e.g. `msw/node` using msw 2 in unit tests for a frontend plugin. Changing the role to `plugin` solves this.

**Comment by freben** [comment]
@grantila So you

- Have an existing frontend plugin
- That plugin has tests that work
- You add `msw@2` to the package devDependencies and add an import of `msw/node` in one of the previously existing tests

And that leads to an issue? What's the error you are seeing?

**Comment by grantila** [comment]
Oh, missed that `customExportConditions`, tried it, doesn't help. In fact, msw infinitely recurses somewhere deep and crashes.

Yeah, we have a `frontend-plugin` with some test files using React and tests the UI itself. Some other test files use `msw@2`, but can't import `msw/node`:

```
Cannot find module 'msw/node' from 'the-unit-test-file.test.ts'
```

Changing the role to `node-library` makes the msw2-tests to work (now it can find `msw/node`), but the UI tests fail: `window not defined`.

Pretty stuck between a rock and a hard place 😬

**Comment by Rugvip** [comment]
It's very strange to me that the package role has any affect on the MSW import tbh, this is how jest config differs across package roles: https://github.com/backstage/backstage/blob/86d6c1d2bfdb90e3ed4712752b2653e5a032d8b5/packages/cli/config/jest.js#L122-L139

I do have to say that https://github.com/mswjs/msw/blob/ca6cb7e8e95ad76c6fe6c44d7f4b80d36bc53478/package.json#L33 is looking interesting, perhaps the `browser` entry somehow gets picked up by Jest if it's using the `jsdom` env?

... and yep that's exactly it, digging into Jest docs: https://jestjs.io/docs/configuration#testenvironmentoptions-object

> Both jest-environment-jsdom and jest-environment-node allow specifying customExportConditions, which allow you to control which versions of a library are loaded from exports in package.json. jest-environment-jsdom defaults to ['browser']. jest-environment-node defaults to ['node', 'node-addons'].

Not sure what the fix is tbh, I'd kinda expect `msw/node` to not define a `browser` entry at all 🤔
